### PR TITLE
CLICKHOUSE-4179 : arrayEnumerateDenseRanked arrayEnumerateUniqRanked 

### PR DESCRIPTION
**Pull request by proller**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):
Two new functions: arrayEnumerateDenseRanked arrayEnumerateUniqRanked

### CODE REVIEW

File: dbms/src/Functions/arrayEnumerateRanked.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    const size_t num_arguments = arguments.size();
+    DepthType clear_depth = 1;
+    DepthType max_array_depth = 1;
+    DepthTypes depths;
+
+    for (size_t i = 0; i < num_arguments; ++i)
+    {
+        if (!arguments[i].column)
+            continue;
+        const auto non_const = arguments[i].column->convertToFullColumnIfConst();
```
Comment (KochetovNicolai) [inline dbms/src/Functions/arrayEnumerateRanked.cpp]: Const column will be converted to full here, but we only need to check it isn't an array. It's better to make `typeid_cast` or something.

File: dbms/src/Functions/arrayEnumerateRanked.h [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+{
+public:
+    static FunctionPtr create(const Context & /* context */) { return std::make_shared<Derived>(); }
+
+    String getName() const override { return Derived::name; }
+
+    bool isVariadic() const override { return true; }
+    size_t getNumberOfArguments() const override { return 0; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply dbms/src/Functions/arrayEnumerateRanked.h]: This function doesn't check that:
* Arguments are valid (arrays or numeric)
* (the number of array arguments) + 1 == (the number of numeric arguments)
* Argument's are in correct order (e.g. number, array, number, array, number; not number, number, number, array, array). Not sure if we need it.

File: dbms/src/Functions/arrayEnumerateRanked.h [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+
+    std::vector<const ColumnArray::Offsets *> offsets_by_depth;
+    std::vector<ColumnPtr> offsetsptr_by_depth; // TODO make one !
+
+    size_t array_num = 0;
+    for (size_t i = 0; i < num_arguments; ++i)
+    {
+        const ColumnPtr & array_ptr = block.getByPosition(arguments[i]).column;
+        const ColumnArray * array = checkAndGetColumn<ColumnArray>(array_ptr.get());
+        if (!array)
```
Comment (KochetovNicolai) [inline dbms/src/Functions/arrayEnumerateRanked.h]: The same code is in `get_array_column`.

File: dbms/src/Functions/arrayEnumerateRanked.h [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+                break;
+
+            if (offsets_by_depth.size() <= col_depth)
+            {
+                offsets_by_depth.emplace_back(&array->getOffsets());
+                offsetsptr_by_depth.emplace_back(array->getOffsetsPtr());
+            }
+            else
+            {
+                if (*offsets_by_depth[col_depth] != array->getOffsets())
```
Comment (KochetovNicolai) [inline dbms/src/Functions/arrayEnumerateRanked.h]: Duplicating code (with line 168)

File: dbms/src/Functions/arrayEnumerateRanked.h [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+*/
+
+    std::vector<size_t> indexes_by_depth(max_array_depth);
+    std::vector<size_t> current_offset_n_by_depth(max_array_depth);
+
+    UInt32 rank = 0;
+
+    for (size_t off : offsets)
+    {
+        std::vector<size_t> indexes(columns.size());
```
Comment (KochetovNicolai) [inline dbms/src/Functions/arrayEnumerateRanked.h]: The name has a kind of conflict with `indices`, which is hash table.

Comment (KochetovNicolai) [reply dbms/src/Functions/arrayEnumerateRanked.h]: Can't we move it outside for loop in order not to allocate memory each time?

File: dbms/src/Functions/arrayEnumerateRanked.h [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+
+            // Debug: DUMP(off, prev_off, j, indexes, res_values[j], columns);
+
+            for (int depth = current_offset_depth - 1; depth >= 0; --depth)
+            {
+                ++indexes_by_depth[depth];
+
+                if (indexes_by_depth[depth] == (*offsets_by_depth[depth])[current_offset_n_by_depth[depth]])
+                {
+                    if (static_cast<int>(clear_depth - 1) == depth)
```
Comment (KochetovNicolai) [inline dbms/src/Functions/arrayEnumerateRanked.h]: Can check that `depth + 1 == clear_depth` to avoid static_cast.

**Review by KochetovNicolai — CHANGES_REQUESTED** [review]

File: dbms/src/Functions/arrayEnumerateRanked.h (hunk 6 above)
Comment (proller) [reply dbms/src/Functions/arrayEnumerateRanked.h]: ../dbms/src/Functions/arrayEnumerateRanked.h:336:37: error: comparison of integers of different signs: 'DB::DepthType' (aka 'unsigned int') and 'int' [-Werror,-Wsign-compare]
                    if (clear_depth == depth + 1)
                        ~~~~~~~~~~~ ^  ~~~~~~~~~

File: dbms/src/Functions/arrayEnumerateRanked.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (proller) [reply dbms/src/Functions/arrayEnumerateRanked.h]: now numeric arguments is optional, cases like ([],1,2,3) now checked

File: dbms/src/Functions/arrayEnumerateRanked.h (hunk 4 above)
Comment (proller) [reply dbms/src/Functions/arrayEnumerateRanked.h]: now its not trivial to remake

**Review by KochetovNicolai — APPROVED** [review]

File: dbms/src/Functions/arrayEnumerateUniqRanked.cpp [hunk 7]
```diff
@@ -0,0 +1,25 @@
+#include "Functions/FunctionFactory.h"
+#include "arrayEnumerateRanked.h"
+
+
+namespace DB
+{
+/** arrayEnumerateUniqRanked(clear_depth, arr1, depth1, arr2, depth2)
```
Comment (alexey-milovidov) [inline dbms/src/Functions/arrayEnumerateUniqRanked.cpp]: Obsolete (and insufficient) comment.

**Pull request merged**